### PR TITLE
PoE genesis restructuring

### PR DESCRIPTION
**Pull request by alpe**
After #348 
Move Genesis PoE contract config for seed mode under `SeedContracts` element.

### CODE REVIEW

File: docs/proto/proto-docs.md
Comment (ethanfrey) [inline docs/proto/proto-docs.md:308]: Since all the "Seed Mode" is moved into it's own key (I agree), I want to check two things:

1. Is `contracts` all we will ever need in non-seed mode? (If not, let's make another sub-structure for it)
2. If this is all we will use, is there a way to rename it, to show this is only for importing existing chains (ignored on initial "clean" genesis)?

File: x/poe/bootstrap.go
Comment (ethanfrey) [inline x/poe/bootstrap.go:64]: :+1: makes it clear what info is used here

File: x/poe/module.go
Comment (ethanfrey) [inline x/poe/module.go:170]: If you explicitly check for seedMode here, do we want an explicit check for non-seed mode (recovering addresses from genesis)?

File: x/poe/module.go
Comment (ethanfrey) [inline x/poe/module.go:177]: We could then avoid passing the entire genesisState to InitGenesis, and just the Params that is used in both cases

File: x/poe/module.go
Comment (ethanfrey) [inline x/poe/module.go:195]: Nice catch. On non-seed mode, there may be existing validators and no diff.

What really needs to be checked however, is if there are any validators after the diff is applied. Maybe there is a better way to check that?

File: x/poe/types/genesis.go
Comment (ethanfrey) [inline x/poe/types/genesis.go:100]: Should we error here if both are set? Or neither is set?

`(g.SeedContracts != nil) == (len(g.Contracts) > 0)`

That would be nice in the validation step at least to avoid some confusion.

File: x/poe/types/genesis.go
Comment (ethanfrey) [inline x/poe/types/genesis.go:103]: That would cover this check and the both set case

File: x/poe/types/genesis.go
Comment (ethanfrey) [inline x/poe/types/genesis.go:116]: Very good check after loading.

Why the -1 btw? We shouldn't be doing any auto-creation on init in this state import phase

File: x/poe/types/genesisio.go
Comment (ethanfrey) [inline x/poe/types/genesisio.go:38]: Nice check here as well!

**Review by ethanfrey — APPROVED** [review]
Good stuff.

I added some minor questions and improvements.
The only thing that really raised my eyebrows was the change from the Operator address to the Tendermint pub key address. Right now, I don't have the context/depth to understand if that is fixing a bug or creating a bug, but it did make me think it deserves a closer look (by someone with the context).

I would approve this for merge pending some minor cleanup and review by the author.

File: app/export.go
Comment (ethanfrey) [reply app/export.go:81]: Update: after reviewing the base PR #349 I agree this is correct

Comment (alpe) [reply app/export.go:81]: ja, fails without this in a manual tests. To be fair, this should have gone into #348 .

File: x/poe/module.go
Comment (alpe) [reply x/poe/module.go:170]: Addresses are checked in `VerifyPoEContracts()`

File: x/poe/module.go
Comment (alpe) [reply x/poe/module.go:195]: Shouldn't tendermint check that? I was assuming some error when no validators are left

File: x/poe/types/genesis.go
Comment (alpe) [reply x/poe/types/genesis.go:100]: yes, this check is some lines above

File: x/poe/types/genesis.go
Comment (alpe) [reply x/poe/types/genesis.go:116]: it is -1 because one element is used for the `UNDEFINED` value, which is not used.

File: docs/proto/proto-docs.md
Comment (alpe) [reply docs/proto/proto-docs.md:308]: A `oneof` element would be cleanest IMHO as it separates the two use cases.  I still need to find out if we need to export historical informations. I have also asked on IBC discord now.

File: x/poe/module.go
Comment (alpe) [reply x/poe/module.go:195]: Found this in cosmos-sdk: https://github.com/cosmos/cosmos-sdk/blob/v0.45.4/baseapp/abci.go#L67-L72

Comment (alpe) [reply x/poe/module.go:195]: It was actually required to return the list of active validators. I am also checking for non empty list.

**Comment by alpe** [comment]
Merging this into `340-state_export_import now` as this PR contains more than just genesis refactorings.

**Pull request merged**
